### PR TITLE
Dev mode

### PR DESCRIPTION
**Pull request by buthanoid**
Automatic finding and crafting of ServiceResults to help testing the GUI in development

### CODE REVIEW

**Comment by buthanoid** [comment]
Maybe add a Log when service guess fails and default to WISARD

**Comment by bourgesl** [comment]
Looks cool !!!
To guess the service we could check 1 specific keyword for each service... ... ....
until we store the service.program as a key

**Comment by buthanoid** [comment]
There is a problem with the current proposition of alternative constructor of ServiceResult :
- it calls this.loadOIFitsFile() and this.loadExecutionLogFile(), whereas after the constructor returns, in DevMode.java we find the call : irModel.addServiceResult(serviceResult) which calls again the two cited methods
- if we remove the two methods from the constructor, we cannot guess the Service anymore, since we have not load the OIFitsFile. So we would need to wait after irModel.addServiceResult(serviceResult) to guess the Service. Problem is, addServiceResult tries to update the GUI, so if service is null, it breaks the update. So maybe we should pick a default service, call addServiceResult, then guess the real service and update the UI again. It looks a bit complicated. But the other way is to duplicate calls to  loadOIFitsFile(), so it is not that good either

Another problem :
- in MainPanel, the jEditorPane displays messages. If we add a ServiceResult only by calling irModel.addServiceResult(serviceResult), it does not get correctly updated and display the message : Missing OIData, please load an OIFits. 
- if we call the method irModel.loadOifitsFile(serviceResult.getOifitsFile()) after crafting the serviceResult, jEditorPane gets correctly updated. But the method irModel.loadOifitsFile() seems to duplicate a bit of irModel.addServiceResult(serviceResult), which triggers additional name conflict resolution dialogs in the UI. So it seems bad too.

What do we do ?

**Comment by buthanoid** [comment]
Now the InputFile is not registered anymore. Only the output and the log files. (So in order to test the dev mode the .jmmc-devmode directory must be filtered from input files).

**Comment by buthanoid** [comment]
`getAbsolutePath().toString()` to clean

**Comment by bourgesl** [comment]
It looks ServiceResult.init() could be useful to initialize default values to rating / comments ... to avoid repeated code in constructors, so constructors are simple.

In craftServiceResult(), you should explicitely load the OIFitsFile() : 
- loadOIFitsFile()
- determine the appropriate Service depending on the future OutputParam.ALGORITHM keyword or now using 1 specific keyword per Service (fuzzy matcher) but it will work
- serviceResult.setService(service)

Another typo: avoid string concatenation in log messages:
logger.info(" a " + a) => logger.info(" a {}", a)

File: src/main/java/fr/jmmc/oimaging/DevMode.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    private static List<ServiceResult> searchForFiles () {
+        List<ServiceResult> listSR = new ArrayList<> ();
+        try {
+            if (FILEDIR.isDirectory()) {
+                File[] dirFiles = FILEDIR.listFiles();
+                Comparator<File> cmp = Comparator.comparing(File::getName, Comparator.naturalOrder());
+                Arrays.sort(dirFiles, cmp);
+                
+                for (int i = 1 ; i < dirFiles.length ; i += 2) {
```
Comment (bourgesl) [inline src/main/java/fr/jmmc/oimaging/DevMode.java:54]: I would prefer only scanning "*.fits" files => build File paths but I agree patterns must be found.
Constants should be added to simplify such file patterns as done in ServiceResult constructor: 
```
this.inputFile = inputFile;
this.oifitsResultFile = FileUtils.getTempFile(inputFilename + ".output.fits");
this.executionLogResultFile = FileUtils.getTempFile(inputFilename + ".log.txt");
```
=> determine inputFileName => other paths can be reconstructed by the current ServiceResult constructor

File: src/main/java/fr/jmmc/oimaging/DevMode.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    /** Craft a ServiceResult that is got not from a run but from files, and add the ServiceResult 
+     * @param sExecutionLogResultFile  must not be null. can be invalid file or protected in reading
+     * @param sOifitsResultFile  must not be null. can be invalid file or protected in reading
+     */
+    private static void craftServiceResult (ServiceResult serviceResult) {
+        
+        IRModel irModel = IRModelManager.getInstance().getIRModel();
+
+        // we set the prefered service. TODO : when we will have a Service keyword or similar we can use it here
+        serviceResult.setService(ServiceList.getPreferedService());
```
Comment (bourgesl) [inline src/main/java/fr/jmmc/oimaging/DevMode.java:90]: Load explicitely the OIFitsFile as before and determine the Service should be done here

File: src/main/java/fr/jmmc/oimaging/DevMode.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    String executionLogResultFileName = dirFiles[i].getAbsolutePath().toString();
+                    
+                    String baseName1 = oiFitsResultFileName.substring(0, oiFitsResultFileName.length() - ".log.txt".length());
+                    String baseName2 = executionLogResultFileName.substring(0, executionLogResultFileName.length() - ".output.fits".length());
+                    
+                    if (baseName1.equals(baseName2)) {
+                        
+                        // the input file is fake but will not be opened so it is ok
+                        ServiceResult sr = new ServiceResult(
+                                new File(baseName1), new File(executionLogResultFileName), new File(oiFitsResultFileName));
```
Comment (bourgesl) [inline src/main/java/fr/jmmc/oimaging/DevMode.java:66]: new File(baseName1) does not exist: it should point to the (output) fits file too ?

Comment (buthanoid) [reply src/main/java/fr/jmmc/oimaging/DevMode.java:66]: As long as the Output file does not contains more than the Input File it is good. But if it adds some Input Value that was not used to construct the Result output there is an inconsistency.
For now I am going to make input point to output

**Comment by bourgesl** [comment]
Looks good, I just fixed code formatting + logs statements

**Pull request merged**